File: packages/snow_draw_core/lib/draw/elements/types/arrow/elbow/elbow_constants.dart
/// Centralized constants for elbow routing + editing.
///
/// These values are intentionally shared so routing and editing maintain the
/// same geometric tolerances and padding behavior.
final class ElbowConstants {
  const ElbowConstants._();

  /// Threshold for treating coordinates as aligned or duplicate.
  static const double dedupThreshold = 1;

  /// Epsilon used when testing segment intersections with obstacle bounds.
  static const intersectionEpsilon = 1e-6;

  /// Minimum Manhattan distance before a stable midpoint elbow is used.
  static const double minArrowLength = 8;

  /// Clamp for route coordinates to prevent runaway values.
  static const double maxPosition = 1000000;

  /// Base padding applied around obstacle bounds.
  static const double basePadding = 42;

  /// Padding used for degenerate obstacle overlaps (creates a small exit box).
  static const double exitPointPadding = 2;

  /// Gap multiplier when no arrowhead is present at a bound endpoint.
  static const double elbowNoArrowheadGapMultiplier = 2;

  /// Side padding to keep routes away from bound element edges.
  static const double elementSidePadding = 8;

  /// Minimum padding when correcting bound endpoint directions during edits.
  static const double directionFixPadding = 12;
}

File: packages/snow_draw_rust_engine/src/draw/elements/types/arrow/elbow/elbow_constants.rs
#![allow(dead_code)]

/// Shared geometry constants for elbow arrow routing and editing.
///
/// Translated from:
/// `packages/snow_draw_engine/lib/draw/elements/types/arrow/elbow/elbow_constants.dart`.
pub struct ElbowConstants;

impl ElbowConstants {
    /// Threshold used when deduplicating nearby points.
    pub const DEDUP_THRESHOLD: f64 = 1.0;

    /// Epsilon used for robust floating-point intersection checks.
    pub const INTERSECTION_EPSILON: f64 = 1e-6;

    /// Minimum arrow length accepted by elbow routing.
    pub const MIN_ARROW_LENGTH: f64 = 8.0;

    /// Clamp bound for very large coordinates.
    pub const MAX_POSITION: f64 = 1_000_000.0;

    /// Baseline padding used in elbow routing calculations.
    pub const BASE_PADDING: f64 = 42.0;

    /// Extra padding around endpoint exits.
    pub const EXIT_POINT_PADDING: f64 = 2.0;

    /// Gap multiplier applied when there is no arrowhead.
    pub const ELBOW_NO_ARROWHEAD_GAP_MULTIPLIER: f64 = 2.0;

    /// Padding from element edges while routing.
    pub const ELEMENT_SIDE_PADDING: f64 = 8.0;

    /// Padding used to stabilize direction fixes.
    pub const DIRECTION_FIX_PADDING: f64 = 12.0;
}
